homeassistant/components/media_player/ziggo_mediabox_xl.py

```diff
@@ -0,0 +1,218 @@
+
+    # Is this a manual configuration?
+    if config.get(CONF_HOST) is not None:
+        host = config.get(CONF_HOST)
+        name = config.get(CONF_NAME)
+    elif discovery_info is not None:
+        host = discovery_info.get('host')
+        name = discovery_info.get('name')
+    else:
+        _LOGGER.warning("Cannot determine device")
+        return
+
+    # Only add a device once, so discovered devices do not override manual
+    # config.
+    hosts = []
+    ip_addr = socket.gethostbyname(host)
+    if ip_addr not in known_devices:
+        sock = socket.socket()
+        try:
+            state = sock.connect_ex((ip_addr, 5900))
+        except socket.error:
+            _LOGGER.error("Couldn't connect to %s", ip_addr)
+        if state == 0:
+            hosts.append(ZiggoMediaboxXLDevice(ip_addr, name))
+            known_devices.add(ip_addr)
+        else:
+            _LOGGER.error("Can't connect to %s", host)
+    else:
+        _LOGGER.warning("Ignoring duplicate Ziggo Mediabox XL %s", host)
+    add_devices(hosts, True)
+
+
+class ZiggoMediaboxXLDevice(MediaPlayerDevice):
+    """Representation of a Ziggo Mediabox XL Device."""
+
+    def __init__(self, host, name):
+        """Initialize the device."""
+        # Generate a configuration for the Samsung library
+        self._host = host
+        self._port = {"state": 62137, "cmd": 5900}
+        self._name = name
+        self._state = STATE_UNKNOWN
+        self._channels = {}
+        self._keys = {
+            "POWER": "E0 00", "OK": "E0 01", "BACK": "E0 02",
+            "CHAN_UP": "E0 06", "CHAN_DOWN": "E0 07",
+            "HELP": "E0 09", "MENU": "E0 0A", "GUIDE": "E0 0B",
+            "INFO": "EO 0E", "TEXT": "E0 0F", "MENU1": "E0 11",
+            "MENU2": "EO 15", "DPAD_UP": "E1 00",
+            "DPAD_DOWN": "E1 01", "DPAD_LEFT": "E1 02",
+            "DPAD_RIGHT": "E1 03", "PAUSE": "E4 00", "STOP": "E4 02",
+            "RECORD": "E4 04", "FWD": "E4 05", "RWD": "E4 07",
+            "MENU3": "E4 07", "ONDEMAND": "EF 28", "DVR": "EF 29",
+            "TV": "EF 2A"}
+        for i in range(10):
+            self._keys["NUM_{}".format(i)] = "E3 {:02d}".format(i)
+        self._fetch_channels()
+
+    def update(self):
+        """Retrieve the state of the device."""
+        # Send an empty key to see if we are still connected
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            if sock.connect_ex((self._host, self._port['state'])) == 0:
+                if self._state != STATE_PAUSED:
+                    self._state = STATE_PLAYING
+            else:
+                self._state = STATE_OFF
+            sock.close()
+        except socket.error:
+            _LOGGER.error("Couldn't fetch state from %s", self._host)
+
+    def send_keys(self, keys):
+        """Send keys to the device and handle exceptions."""
+        try:
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
These socket level details need to be moved to a separate python lib published on PyPI. We can provide more guidance in chat if you have more questions here.